datanode/src/uss_metadata.py

```diff
@@ -248,13 +286,44 @@ def remove_operation(self, uss_id, gufi):
         ]
         found = (len(oper['operations']) == num_operations - 1)
         oper['version'] = self.version
-        oper['timestamp'] = self.format_ts()
+        oper['timestamp'] = format_utils.format_ts()
     return found
 
-  def format_ts(self, timestamp=None):
-    r = datetime.datetime.now(pytz.utc) if timestamp is None else timestamp
-    r = r.astimezone(pytz.utc)
-    return '{0}Z'.format(r.strftime('%Y-%m-%dT%H:%M:%S.%f')[:23])
+  def insert_uvr(self, uvr):
+    """Inserts a UVR, failing if another UVR with the same ID already exists.
+
+    Args:
+      uvr: UVR instance
+
+    Raises:
+      ValueError: if a UVR already exists with the new UVR's message_id
+    """
+
+    if uvr['message_id'] in [existing['message_id'] for existing in self.uvrs]:
```

Comment:
This behavior is different from girdcell operator or operation, we overwrite with PUT (TCL4 didn't want to support PATCH). Can we do the same here?

Reply:
I don't think we want to do that because part of the UVR itself is its coordinate outline.  What happens if a client tries to update a UVR in grid cells (0,0) and (0,1) with a new UVR of the same ID in grid cells (0,0) and (1,0)?  Currently, the client only needs to send the UVR itself and coordinates are extracted from it; with an update, we could no longer extract coordinates for the old UVR to be replaced from the update request so we would need to add an additional coordinates input parameter.  But, then that parameter could conflict with the UVR itself, and what should happen in that case?  There probably shouldn't be an error because that would be inconsistent with the behavior during update where error checking cannot be guaranteed.  But it also seems very odd to allow clients to update a UVR when we know coordinates specifications don't match.  I think the exact-UVR-match approach is a clean way to ensure UVR consistency in the grid while adding little to no overhead to what the client would already need to do to delete UVRs (and perhaps less overhead since full UVR specification eliminates the need to specify coordinates or grid cells separately).

Reply:
Roger. We'll address it specifically as we support PATCH, as that should do the changes to the grid if the boundary changes.